django/db/models/query.py

```diff
@@ -1618,14 +1622,44 @@ def all(self):
         """
         return self._chain()
 
-    def filter(self, *args, **kwargs):
+    def _class_filter(cls, *args, **kwargs):
+        if invalid_kwargs := PROHIBITED_FILTER_KWARGS.intersection(kwargs):
+            invalid_kwargs_str = ", ".join(f"'{k}'" for k in sorted(invalid_kwargs))
+            raise TypeError(f"The following kwargs are invalid: {invalid_kwargs_str}")
+        initial_filter = Q(*args, **kwargs)
+        # Chain initial filters.
+        if cls._initial_filter is not None:
+            initial_filter = cls._initial_filter & initial_filter
+            base_class = cls.__bases__[0]
+        else:
+            base_class = cls
+
+        initial_filter_id = id(initial_filter)
+        name = f"{base_class.__name__}WithFilter{initial_filter_id}"
+        new_filter_class = type(
+            name,
+            (base_class,),
+            {
+                "_initial_filter": initial_filter,
+                "__module__": base_class.__module__,
+                "__qualname__": f"{base_class.__qualname__}.{name}",
+            },
+        )
+        # Add new class to the QuerySet class to make it pickleable.
+        setattr(base_class, name, new_filter_class)
```

Comment:
Unfortunately this approach won't work if a pickle is shared between interpreters (which is common for queryset when they are cached) because

1. `id` will return distinct values
2. `Q` instances might be gargabe collected
3. Queryset with required filter types might not be created at the time of unpickling

What we really need here is a `__reduce__` [solution](https://docs.python.org/3/library/pickle.html#object.__reduce__) that returns something like `(QuerySet._create_with_filter, (base_class, initial_filter), {})` and memoize from `Q.identity` instead of `id`.